### PR TITLE
docs(operators): Adds instructions for adding Cluster Operator replicas

### PR DESCRIPTION
**Pull request by PaulRMellor**
Signed-off-by: prmellor <pmellor@redhat.com>

**Documentation**
Adds a new assembly with concepts and a procedure for running multiple Cluster Operator replicas with leader election.
Also adds a reference to this new functionality in the Deploying and Overview guides.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: documentation/assemblies/deploying/assembly-deploy-cluster-operator.adoc [hunk 1]
```diff
@@ -10,6 +10,8 @@ The Cluster Operator is responsible for deploying and managing Kafka clusters wi
 
 When the Cluster Operator is running, it starts to watch for updates of Kafka resources.
 
+By default, a single Cluster Operator is deployed. You can add replicas with leader election so that additional Cluster Operators are on standby in case of disruption. For more information, see link:{BookURLUsing}#assembly-using-multiple-cluster-operator-replicas-str[Running multiple Cluster Operator replicas with leader election].
```
Comment (scholzj) [inline documentation/assemblies/deploying/assembly-deploy-cluster-operator.adoc:13]:
```suggestion
By default, a single replica of the Cluster Operator is deployed. You can add replicas with leader election so that additional Cluster Operators are on standby in case of disruption. For more information, see link:{BookURLUsing}#assembly-using-multiple-cluster-operator-replicas-str[Running multiple Cluster Operator replicas with leader election]. 
```

File: documentation/assemblies/operators/assembly-using-multiple-cluster-operators.adoc [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// This assembly is included in the following assemblies:
+//
+// assembly-using-the-cluster-operator.adoc
+
+[id='assembly-using-multiple-cluster-operator-replicas-{context}']
+= Running multiple Cluster Operator replicas with leader election
+
+[role="_abstract"]
+Enable xref:ref-operator-cluster-leader-election-{context}[leader election] in the Cluster Operator configuration to run multiple parallel replicas of the Cluster Operator.
```
Comment (scholzj) [inline documentation/assemblies/operators/assembly-using-multiple-cluster-operators.adoc:9]: Should we add that its enabled by default in the installation files?

File: documentation/assemblies/operators/assembly-using-multiple-cluster-operators.adoc [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+One replica is elected as the active leader and operates the deployed resources.
+The other replicas run in standby mode.
+When the leader stops or fails, one of the standby replicas is elected as the new leader and starts operating the deployed resources.
+
+By default, Strimzi runs with a single Cluster Operator that is always the leader replica.
```
Comment (scholzj) [inline documentation/assemblies/operators/assembly-using-multiple-cluster-operators.adoc:14]:
```suggestion
By default, Strimzi runs with a single Cluster Operator replica that is always the leader replica. 
```

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+[source,shell,subs="+quotes"]
+----
+kubectl get deployments -n myproject
+----
++
+.Output shows the deployment name and readiness
+[source,shell,subs="+quotes"]
+----
+NAME                      READY  UP-TO-DATE  AVAILABLE
+strimzi-cluster-operator  1/1    1           3
```
Comment (scholzj) [inline documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:117]: This is how it should look like with 3 replicas when they are all ready:

```suggestion
strimzi-cluster-operator  3/3    3           3
```

File: documentation/modules/operators/ref-operator-cluster.adoc [hunk 5]
```diff
@@ -217,17 +217,21 @@ Enables or disables features and functionality controlled by xref:ref-operator-c
...
-Enables or disables the leader election which allows to run additional operator instances as a hot-stand-by.
-Disabled by default.
+[id='ref-operator-cluster-leader-election-{context}']
+== Leader election environment variables 
+
+Use leader election environment variables when xref:assembly-using-multiple-cluster-operator-replicas-{context}[running additional Cluster Operator replicas].
+You might run additional replicas to safeguard against disruption caused by major failure.
+
+`STRIMZI_LEADER_ELECTION_ENABLED`:: Optional, disabled (`false`) by default.
+Enables or disables leader election that allows additional Cluster Operator replicas to run on standby.
```
Comment (scholzj) [inline documentation/modules/operators/ref-operator-cluster.adoc:227]: Should we clarify that it is disabled by default when the environment variable is set, but that the installation files actually enable it?

File: documentation/modules/overview/con-overview-components-cluster-operator.adoc [hunk 6]
```diff
@@ -6,24 +6,29 @@
 [id='overview-components-cluster-operator-{context}']
 = Cluster Operator
 
-Strimzi uses the Cluster Operator to deploy and manage clusters for:
+[role="_abstract"]
+Strimzi uses the Cluster Operator to deploy and manage clusters.
+By default, when you deploy Strimzi a single Cluster Operator is deployed.
```
Comment (scholzj) [inline documentation/modules/overview/con-overview-components-cluster-operator.adoc:11]:
```suggestion
By default, when you deploy Strimzi a single Cluster Operator replica is deployed. 
```

**Review by scholzj — COMMENTED** [review]
I left some comments. I think it is important to distinguish two situations:
* Having a single installation of the Cluster Operator with one or more replicas
* Having multiple installations of Cluster Operator

The leader election applies to the first one. That is why on some places I commented on the working of _single Cluster Operator_ to put more focus on the replicas.

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc [hunk 7]
```diff
@@ -0,0 +1,122 @@
+// This assembly is included in the following assemblies:
+//
+// assembly-using-multiple-cluster-operators.adoc
+
+[id='assembly-configuring-proxy-config-cluster-operator-{context}']
+= Configuring Cluster Operator replicas
+
+[role="_abstract"]
+Increase the number of replicas and enable leader election to run additional Cluster Operator replicas in standby mode.
```
Comment (tomncooper) [inline documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:9]:
```suggestion
To run additional Cluster Operator replicas in standby mode, you will need to increase the number of replicas and enable leader election.
```

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+To configure leader election, use leader election environment variables.
```
Comment (tomncooper) [inline documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:10]:
```suggestion
To configure leader election, use the leader election environment variables.
```

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+To make the required changes, you can configure the following Cluster Operator installation files located in `install/cluster-operator/`:
```
Comment (tomncooper) [inline documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:12]:
```suggestion
To make the required changes, you should configure the following Cluster Operator installation files located in `install/cluster-operator/`: 
```

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+Increase the number of replicas and enable leader election to run additional Cluster Operator replicas in standby mode.
+To configure leader election, use leader election environment variables.
+
+To make the required changes, you can configure the following Cluster Operator installation files located in `install/cluster-operator/`: 
+
+* 060-Deployment-strimzi-cluster-operator.yaml
+* 022-ClusterRole-strimzi-cluster-operator-role.yaml
+* 022-RoleBinding-strimzi-cluster-operator.yaml
+
+Leader election has its own its own `ClusterRole` and `RoleBinding` RBAC resources that target the namespace where Cluster Operator is running, rather than the namespace it is watching.
```
Comment (tomncooper) [inline documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:18]:
```suggestion
Leader election has its own `ClusterRole` and `RoleBinding` RBAC resources that target the namespace where the Cluster Operator is running, rather than the namespace it is watching. 
```

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+If you use a different `Lease` name or namespace, update the `ClusterRole` and `RoleBinding` files accordingly.
+
+.Prerequisites
+
+* This procedure requires use of a Kubernetes user account which is able to create `CustomResourceDefinition`, `ClusterRole`, and `ClusterRoleBinding` resources.
+Use of Role Base Access Control (RBAC) in the Kubernetes cluster usually means that permission to create, edit, and delete these resources is limited to Kubernetes cluster administrators, such as `system:admin`.
+
+.Procedure
+
+Edit the `Deployment` resource used to deploy the Cluster Operator in the `060-Deployment-strimzi-cluster-operator.yaml` file.
```
Comment (tomncooper) [inline documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:32]:
```suggestion
Edit the `Deployment` resource which is used to deploy the Cluster Operator and is defined in the `060-Deployment-strimzi-cluster-operator.yaml` file.
```

File: documentation/assemblies/operators/assembly-using-multiple-cluster-operators.adoc (hunk 2 above)
Comment (PaulRMellor) [reply documentation/assemblies/operators/assembly-using-multiple-cluster-operators.adoc:9]: Changed to:

> The default Cluster Operator configuration enables [leader election]. 
> Use leader election to run multiple parallel replicas of the Cluster Operator.

File: documentation/modules/operators/ref-operator-cluster.adoc (hunk 5 above)
Comment (PaulRMellor) [reply documentation/modules/operators/ref-operator-cluster.adoc:227]: I've added this note. Not sure if it's clear enough:

> NOTE: Leader election is disabled by default.
> It is only enabled when applying this environment variable on installation.

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
++ 
+[source,yaml,numbered,options="nowrap",highlight='12']
+.Updating the RoleBinding references to the lease 
+----
+include::../install/cluster-operator/022-RoleBinding-strimzi-cluster-operator.yaml[lines=1..8]
+    name: my-strimzi-cluster-operator
+    namespace: myproject
+# ...
+----
+
```
Comment (scholzj) [inline documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:100]: Since this is all done by default ... should we formulate it more as _Doublecheck that these are set and if not set them?

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+The default deployment configuration creates a `Lease` resource called `strimzi-cluster-operator` in the same namespace as the Cluster Operator.
+The Cluster Operator uses leases to manage leader election. 
+The RBAC resources provide the permissions to use the `Lease` resource.
+If you use a different `Lease` name or namespace, update the `ClusterRole` and `RoleBinding` files accordingly.
+
+.Prerequisites
+
+* This procedure requires use of a Kubernetes user account which is able to create `CustomResourceDefinition`, `ClusterRole`, and `ClusterRoleBinding` resources.
+Use of Role Base Access Control (RBAC) in the Kubernetes cluster usually means that permission to create, edit, and delete these resources is limited to Kubernetes cluster administrators, such as `system:admin`.
```
Comment (scholzj) [inline documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:28]: I would probably rephrase it to _you need account with privileged to manage these resources. I would leave the cluster or system admin our of it.

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc (hunk 12 above)
Comment (PaulRMellor) [reply documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:100]: Okay. I changed to:

> . Check that the leader election `env` properties are set.
> +
> If they are not set, configure them.

File: documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc (hunk 13 above)
Comment (PaulRMellor) [reply documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:28]: Changed to:

> You need an account with permission to create and manage `CustomResourceDefinition` and RBAC (`ClusterRole`, and `RoleBinding`) resources.

Comment (PaulRMellor) [reply documentation/modules/operators/proc-configuring-multiple-cluster-operators.adoc:28]: @scholzj -- We use the same prereq in the Deploying guide for deploying the Cluster Operator. Should we reduce down there too? I don't think we lose anything

**Review by scholzj — APPROVED** [review]
LGTM. Thanks.

File: documentation/modules/operators/ref-operator-cluster.adoc [hunk 14]
```diff
@@ -217,17 +217,24 @@ Enables or disables features and functionality controlled by xref:ref-operator-c
...
 Configures the pluggable `PodSecurityProvider` class which can be used to provide the security context configuration for Pods and containers.
 
-[[STRIMZI_LEADER_ELECTION_ENABLED]]
-`STRIMZI_LEADER_ELECTION_ENABLED`:: Optional, default `false`.
-Enables or disables the leader election which allows to run additional operator instances as a hot-stand-by.
-Disabled by default.
+[id='ref-operator-cluster-leader-election-{context}']
+== Leader election environment variables 
+
+Use leader election environment variables when xref:assembly-using-multiple-cluster-operator-replicas-{context}[running additional Cluster Operator replicas].
```
Comment (tomncooper) [inline documentation/modules/operators/ref-operator-cluster.adoc:223]:
```suggestion
Use the leader election environment variable when xref:assembly-using-multiple-cluster-operator-replicas-{context}[running additional Cluster Operator replicas].
```

**Review by tomncooper — APPROVED** [review]
